### PR TITLE
Add topology

### PR DESCRIPTION
**Pull request by Eyusd**
Some functions to access registers have been added in Bus.cc
Tools include new files to check slaves properties and deduce network topology. It is exported in .csv file.

The python script reads such a csv and print the network shape (slaves are identified by their address)

Example of result :
image

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,92 @@
+{
```
Comment (Siviuze) [inline .vscode/settings.json:1]: vscode project files mixed with this PR: to remove (feel free to update .gitignore)

File: examples/easycat/easycat_example.cc [hunk 2]
```diff
@@ -155,4 +155,4 @@ int main(int argc, char* argv[])
     fclose(stat_file);
 
     return 0;
-}
+}
```
Comment (Siviuze) [inline examples/easycat/easycat_example.cc:158]: unrelated modification

File: include/kickcat/Bus.h [hunk 3]
```diff
@@ -6,6 +6,7 @@
 #include <list>
 #include <vector>
 #include <functional>
+#include <map>
```
Comment (Siviuze) [inline include/kickcat/Bus.h:9]: Prefer unordered_map unless you need to loop on the container with a sorted behavior.

File: src/Bus.cc [hunk 4]
```diff
@@ -933,4 +934,101 @@ namespace kickcat
...
+        std::stack<uint16_t> branches;
+        for (auto& slave : slaves_) 
+        {   
+            int openPorts = slave.countOpenPorts();
+
+            if (openPorts == 1)
+            {
+                topology_[slave.address] = lastSeen;
+                lastSeen = slave.address;
+                if (!branches.empty())
```
Comment (Siviuze) [inline src/Bus.cc:950]: Prefer not over !

File: src/Bus.cc [hunk 5]
```diff
@@ -933,4 +934,101 @@ namespace kickcat
...
+
+    void Bus::getTopology()
+    {
+        uint16_t lastSeen = slaves_.at(0).address;
+        std::stack<uint16_t> branches;
+        for (auto& slave : slaves_) 
+        {   
+            int openPorts = slave.countOpenPorts();
+
+            if (openPorts == 1)
```
Comment (Siviuze) [inline src/Bus.cc:946]: switch case?

File: src/Bus.cc [hunk 6]
```diff
@@ -933,4 +934,101 @@ namespace kickcat
...
+            else if (openPorts == 2)
+            {
+                topology_[slave.address] = lastSeen;
+                lastSeen = slave.address;
+            }
+            else 
+            {
+                topology_[slave.address] = lastSeen;
+                lastSeen = slave.address;
+                for (int i = 2; i < openPorts/2; i++)
```
Comment (Siviuze) [inline src/Bus.cc:965]: Prefer ++i

File: src/Bus.cc [hunk 7]
```diff
@@ -933,4 +934,101 @@ namespace kickcat
...
+    }
+
+    void Bus::sendGetRegister(Slave& slave, uint16_t reg)
+    {
+        slave.reg_to_check = 0;
+        auto process = [&slave](DatagramHeader const*, uint8_t const* data, uint16_t wkc)
+        {
+            if (wkc != 1)
+            {
+                return DatagramState::OK;
```
Comment (Siviuze) [inline src/Bus.cc:996]: OK feedback with a wrong working counter

File: src/Bus.cc [hunk 8]
```diff
@@ -933,4 +934,101 @@ namespace kickcat
...
+        link_.processDatagrams();
+    }
+
+    void Bus::sendGetDLStatus(Slave& slave)
+    {
+        auto process = [&slave](DatagramHeader const*, uint8_t const* data, uint16_t wkc)
+        {
+            if (wkc != 1)
+            {
+                return DatagramState::OK;
```
Comment (Siviuze) [inline src/Bus.cc:1018]: OK feedback with a wrong working counter

File: src/Bus.cc [hunk 9]
```diff
@@ -933,4 +934,101 @@ namespace kickcat
...
+                printf( "Slave %04x parent : slave %04x \n", slave.address, topology_[slave.address]);
+            }
+            else
+            {
+                printf( "Slave %04x parent : master \n", slave.address);
+            }
+        }
+    }
+
+    void Bus::sendGetRegister(Slave& slave, uint16_t reg)
```
Comment (Siviuze) [inline src/Bus.cc:989]: Are all registers guaranteed to be 16bits large? If no a template may be a good option.

File: src/Slave.cc [hunk 10]
```diff
@@ -5,6 +5,29 @@
 
 namespace kickcat
 {
+    void printGeneralEntry(eeprom::GeneralEntry const* ge) {
```
Comment (Siviuze) [inline src/Slave.cc:8]: linefeed and variable name is not meaningful (prefer general_entry or at least entry)

File: src/Slave.cc [hunk 11]
```diff
@@ -263,4 +286,38 @@ namespace kickcat
 
         return sum;
     }
+
+    int Slave::countOpenPorts() 
+    {
+        return dl_status.PL_port0 +
+               dl_status.PL_port1 +
+                dl_status.PL_port2 +
```
Comment (Siviuze) [inline src/Slave.cc:294]: align

File: tools/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,14 @@
+set(EXAMPLE_SRCS topology.cc)
+if (PIKEOS)
```
Comment (Siviuze) [inline tools/CMakeLists.txt:2]: Do not build tool for PikeOS for now (not really usable as-is anyway on this embedded OS)

File: tools/displayTopology.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
+import matplotlib.pyplot as plt
```
Comment (Siviuze) [inline tools/displayTopology.py:1]: Add a subfolder for this tool: more tools are coming :)

File: tools/topology.cc [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+        fwrite(sample_str.data(), 1, sample_str.size(), top_file);
+        fwrite(",", 1, 1, top_file);
+        sample_str = std::to_string(value);
+        fwrite(sample_str.data(), 1, sample_str.size(), top_file);
+        fwrite("\n", 1, 1, top_file);
+
+    }
+    fclose(top_file);
+    printf("File written\n");
+}
```
Comment (Siviuze) [inline tools/topology.cc:119]: no end of line

File: tools/topology.cc [hunk 15]
```diff
@@ -0,0 +1,119 @@
+#include "kickcat/Bus.h"
```
Comment (Siviuze) [inline tools/topology.cc:1]: rename this file to match what it is actually doing (scanTopology?)

File: include/kickcat/Bus.h [hunk 16]
```diff
@@ -28,7 +29,7 @@ namespace kickcat
 
         // set the bus from an unknown state to PREOP state
         // 0ms disables the watchdog
-        void init(nanoseconds watchdog = 100ms);
+        void init(nanoseconds watchdog = 0ms);
```
Comment (Rdk-T) [inline include/kickcat/Bus.h:32]: Remove this debug change, unrelated to the PR.

File: include/kickcat/Bus.h [hunk 17]
```diff
@@ -149,6 +158,8 @@ namespace kickcat
 
         nanoseconds tiny_wait{200us};
         nanoseconds big_wait{10ms};
+
+        std::map<uint16_t, uint16_t> topology_;
```
Comment (Rdk-T) [inline include/kickcat/Bus.h:162]: Add a short comment to describe what is associated here (parent id associated to slave id)

File: tools/diagnosticsHelpers/displayTopology.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+parser = argparse.ArgumentParser()
+   
+parser.add_argument('--file', type=str, required=False, help='Path to the network topology - .csv file')
+
+args = parser.parse_args()
+
+if (args.file is None):
+    
+    #We get file from expect path, relative to this one
+    sub_path = os.path.dirname(os.path.realpath(__file__))
```
Comment (Rdk-T) [inline tools/diagnosticsHelpers/displayTopology.py:17]: don't try to find the absolute path, by default use only the name of the file expecting it to be in the current directory and display a message in case the file not exist.

It's better to keep the code simple than to try to guess the needs of the user.

File: tools/diagnosticsHelpers/displayTopology.py [hunk 19]
```diff
@@ -0,0 +1,48 @@
+import matplotlib.pyplot as plt
+import networkx as nx
+import scipy
```
Comment (Rdk-T) [inline tools/diagnosticsHelpers/displayTopology.py:3]: unused ?

File: tools/scanTopology.cc [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+        fwrite(sample_str.data(), 1, sample_str.size(), top_file);
+        fwrite(",", 1, 1, top_file);
+        sample_str = std::to_string(value);
+        fwrite(sample_str.data(), 1, sample_str.size(), top_file);
+        fwrite("\n", 1, 1, top_file);
+
+    }
+    fclose(top_file);
+    printf("File written");
+}
```
Comment (Rdk-T) [inline tools/scanTopology.cc:119]: Missing a newline at the end of the file, have a look to this answer

https://unix.stackexchange.com/a/18789

File: include/kickcat/Bus.h [hunk 21]
```diff
@@ -55,6 +56,9 @@ namespace kickcat
         // It enable users to do one or multiple operations in a row, process something, and process all awaiting frames.
         void sendGetALStatus(Slave& slave, std::function<void(DatagramState const&)> const& error);
 
+        void sendGetRegister(Slave& slave, uint16_t reg);
```
Comment (Siviuze) [inline include/kickcat/Bus.h:59]: No one is using this call?

File: include/kickcat/Slave.h [hunk 22]
```diff
@@ -48,6 +52,10 @@ namespace kickcat
         uint32_t eeprom_size; // in bytes
         uint16_t eeprom_version;
 
+        uint16_t reg_to_check; // For now we only need to check registers that are 16 bits long, this may change in the future
```
Comment (Siviuze) [inline include/kickcat/Slave.h:55]: This member seems useless?

File: include/kickcat/Bus.h [hunk 23]
```diff
@@ -54,6 +54,30 @@ namespace kickcat
         // asynchrone read/write/mailbox/state methods
         // It enable users to do one or multiple operations in a row, process something, and process all awaiting frames.
         void sendGetALStatus(Slave& slave, std::function<void(DatagramState const&)> const& error);
+        void sendGetDLStatus(Slave& slave);
+
+        template<typename T>
+        void sendGetRegister(uint16_t slave_address, uint16_t reg_address, T& value_read)
```
Comment (Rdk-T) [inline include/kickcat/Bus.h:60]: To remove, will be done in a further PR.

File: tools/scanTopology.cc [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+
+    uint8_t io_buffer[2048];
+    try
+    {
+        socket->open(argv[1], 2ms);
+        bus.init();
+
+        print_current_state();
+
+        bus.createMapping(io_buffer);
```
Comment (Rdk-T) [inline tools/scanTopology.cc:48]: Since in this example we only need to access to the DLstatus I don't think we need more than the bus init to be able to access the data.

File: unit/bus-t.cc [hunk 25]
```diff
@@ -1,5 +1,6 @@
 #include <gtest/gtest.h>
 #include <cstring>
+#include <unordered_map>
```
Comment (Rdk-T) [inline unit/bus-t.cc:3]: removed the include since the test has moved.

File: unit/diagnostics-t.cc [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+
+    // Case 4 : invalid topology [lone slave] ( 0 - 1 - 2 - 3    4 )
+
+    slaves[4].dl_status.PL_port0 = 0;
+    slaves[4].dl_status.PL_port1 = 0;
+    slaves[4].dl_status.PL_port2 = 0;
+    slaves[4].dl_status.PL_port3 = 0;
+
+    ASSERT_THROW(getTopology(slaves), Error);
+}
```
Comment (Rdk-T) [inline unit/diagnostics-t.cc:139]: newline

File: unit/slave-t.cc [hunk 27]
```diff
@@ -110,3 +110,25 @@ TEST(Slave, parse_SII)
...
+    for (uint8_t n = 0; n < 16; ++n)
+    {   
+        slave.dl_status.PL_port0 = (n & mask0);
+        slave.dl_status.PL_port1 = (n & mask1) >> 1;
+        slave.dl_status.PL_port2 = (n & mask2) >> 2;
+        slave.dl_status.PL_port3 = (n & mask3) >> 3;
+
+        ASSERT_EQ((n & mask0) + ((n & mask1) >> 1) + ((n & mask2) >> 2) + ((n & mask3) >> 3), slave.countOpenPorts());
+    }
+}
```
Comment (Rdk-T) [inline unit/slave-t.cc:134]: newline

File: .gitignore [hunk 28]
```diff
@@ -1 +1,3 @@
 build/
+.github
```
Comment (Siviuze) [inline .gitignore:2]: TO remove: this is where github CI pipeline is defined.

File: src/Bus.cc [hunk 29]
```diff
@@ -1,4 +1,5 @@
 #include <cstring>
+#include <stack>
```
Comment (Siviuze) [inline src/Bus.cc:2]: Useless in this file

File: src/Slave.cc [hunk 30]
```diff
@@ -5,6 +5,29 @@
 
 namespace kickcat
 {
+    void printGeneralEntry(eeprom::GeneralEntry const* general_entry) {
```
Comment (Siviuze) [inline src/Slave.cc:8]: linefeed before {

**Review by Siviuze — APPROVED** [review]
